### DISCUSSION TITLE
LDF file of Elsa server database grows rapidly

### DISCUSSION THREAD

**Discussion by tsoG2000** [elsa-workflows/elsa-core#1619] [Q&A] [answered]
LDF file of Elsa server database grows rapidly with few stress tests that fail, so the numbers look like: 1 workflow is 1Mb of .ldf file. We were at about 30000 workflows, and .ldf grew to 30Gb. 

What are possible reasons?

Now I found out that even after successful tests that have a lot of async request .ldf file grows at the same rate, this looks like a serious problem. :(

**Comment by sfmskywalker** [comment, marked as answer]
> LDF file of Elsa server database grows rapidly with few stress tests that fail, so the numbers look like: 1 workflow is 1Mb of .ldf file. We were at about 30000 workflows, and .ldf grew to 30Gb.
> 
> What are possible reasons?

The main reason is that the `Elsa.WorkflowInstances` and `Elsa.WorkflowExecutionLog` tables that grow in size during each workflow run.
If you don't want this to happen, there are three options:

1. Use the new **Retention** feature (available with the current 2.4 preview) to periodically delete workflow instances (and their execution log history).
2. Create a custom job or even external scheduled (Windows/Linux) task that performs cleanup. This would also allow you to move workflow instances & journal entries to an archive if you don't want to permanently delete workflow history.
3. Use the In-Memory provider for the workflow instance and the workflow execution log stores if you don't care at all about keeping workflow history.

> Now I found out that even after successful tests that have a lot of async request .ldf file grows at the same rate, this looks like a serious problem. :(

Why wouldn't the .ldf grow, regardless of requests being async or not? Workflow history will be created regardless of the type of requests made to execute workflows.

**Reply by sfmskywalker** [reply]
PS. To setup the Retention feature, you can use the following code in your Startup class:

```csharp
services.AddRetentionServices(options => 
{
   options.SweepInterval = TimeSpan.FromHours(24); // How often the job should query the database for workflows to remove.
   options.TimeToLive = Duration.FromDays(7); // How long workflow instances are allowed to exist in the database before they get removed. Note: only finished, cancelled and faulted workflows will be removed. Suspended workflows will remain in the database.
   options.PageSize = 1000; // The number of workflows to process at the same time. This prevents the server from loading ALL expired workflow instances into memory.
});
```

**Reply by wfurney** [reply]
> PS. To setup the Retention feature, you can use the following code in your Startup class:
> 
> ```cs
> services.AddRetentionServices(options => 
> {
>    options.SweepInterval = TimeSpan.FromHours(24); // How often the job should query the database for workflows to remove.
>    options.TimeToLive = Duration.FromDays(7); // How long workflow instances are allowed to exist in the database before they get removed. Note: only finished, cancelled and faulted workflows will be removed. Suspended workflows will remain in the database.
>    options.PageSize = 1000; // The number of workflows to process at the same time. This prevents the server from loading ALL expired workflow instances into memory.
> });
> ```

Is it possible that this feature was broken in a recent release? I have been seeing this exception since a recent upgrade: 

Elsa.Retention.HostedServices.CleanupHostedService.ExecuteAsync(CancellationToken stoppingToken)
  Exception data:
    Severity: ERROR
    SqlState: 42703
    MessageText: column "workflowinstanceid" does not exist
    Hint: Perhaps you meant to reference the column "WorkflowExecutionLogRecords.WorkflowInstanceId".
    Position: 56
    File: parse_relation.c
    Line: 3551
    Routine: errorMissingColumn

I haven't had time to do a deeper investigation yet.
